### PR TITLE
Release/hip 820. Added 3 methods

### LINKED ISSUE #14: hedera_signQueryAndSend
**Describe the spec as defined in HIP-820**
When a dApp needs the services of the controller to perform an Hedera Network Query, it can use the hedera_signQueryAndSend method. This method accepts a base64 encoded protobuf representation of the Hedera API [Query](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/query.proto#L68) message as input. The controller decodes the message, interprets the contents, and if valid and approved, adds an appropriate signed transaction to the header field of the [Query](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/query.proto#L68) (if applicable), repackages the message as appropriate and sends the message to the network. The controller shall wait for the query response and returns the resulting base64 encoded protobuf Response message received from the network.

The controller is allowed to retry submitting the query to the Hedera Network for certain types of error conditions, such as gRPC connection droppage or Hedera BUSY signals. This specification does not obligate the controller to retry, but recommends the practice to help improve perceived user experience for consumer applications.

Parameters
query – a base64 encoding of the protobuf encoded Hedera API [Query](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/query.proto#L68) message. The message may or may not contain a preliminary [QueryHeader](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/query_header.proto#L66) value in the header field of the enclosed typed query message. It is the responsibility of the controller to complete the necessary information contained in the header field prior to submission to the network.

Note: the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) Scope property of the RPC call shall be used to identify which network shall be assumed when signing and sending the query if multiple chainids were authorized in the pairing process. It is not necessary to transmit that value as a method parameter.

Returns
response – a base64 encoding of the protobuf encoded Hedera API [Response](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/response.proto#L68) message. The response is passed back to the caller without modification and may or may not indicate success. It is the responsibility of the dApp to decode this protobuf representation into the expected query response results.

**Does this issue add new functionality to this repository or change functionality (such as a change in function signature)**
This is a new JSON-RPC function, but will use some of the other pieces of this library.

### LINKED ISSUE #15: hedera_signMessage
**Describe the spec as defined in HIP-820**
When a dApp needs to challenge the controller for an off-ledger authentication such as a website login, it can use the hedera_signMessage method. This method accepts a plain text string value as input. If approved by the user, the controller computes the keccak256 hash of the UTF-8 encoded message prepended with "\x19Hedera Signed Message:\n" plus the length of the message, then signs the resulting hash value. The resulting signature(s) is transmitted back to the user encoded in a SignatureMap structure. The pseudo code for computing the signature is as follows:

sign(keccak256("\x19Hedera Signed Message:\n" + len(message) + message))
It is necessary to add the prefix to the message prior to signing to prevent misuse of the method where a bad actor could request a signature from the controller that could be executable on the hedera network or other chain (in the case where the controller is a multi-chain controller).

Parameters
message – a plain text string to present to the user prior to authorizing a signature. It is recommended that this be a human readable string and contain identifying information including a timestamp and/or other transient information that can be validated by the user to help prevent replay attacks.

Note: the [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) Scope property of the RPC call shall be used to identify which network shall be assumed when signing the message if multiple chainids were authorized in the pairing process. It is not necessary to transmit that value as a method parameter.

Returns
signatureMap – a base64 encoding of the protobuf endoded Hedera API [SignatureMap](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/basic_types.proto#L780) message. The encoded structure must include at least one signature within the property’s [SignatureMap](https://github.com/hashgraph/hedera-protobufs/blob/f36e05bd6bf3f572707ca9bb338f5ad6421a4241/services/basic_types.proto#L780) structure. It is allowed to provide more if the context warrants multiple signatures.

**Does this issue add new functionality to this repository or change functionality (such as a change in function signature)**
This is currently a function in this library, though the spec lays out an additional security measure by adding a prefix to the message makes the calculated signature recognizable as an Ethereum specific signature. This prevents misuse where a malicious DApp can sign arbitrary data (e.g. transaction) and use the signature to impersonate the victim.

This is replicated from https://specs.walletconnect.com/2.0/blockchain-rpc/ethereum-rpc

**Additional context**
Are there other options that should be explored while implementing this issue?

### LINKED ISSUE #16: hedera_getNodeAddresses
**Describe the spec as defined in HIP-820**
While constructing a transaction for transmission to a controller, a dApp needs to choose which Hedera Network node shall receive the transaction prior to signing (this is a requirement of the Hedera API Protocol). While a dApp can easily obtain a list of potential Hedera Nodes, a controller may not have an all-inclusive list nor a path to the node’s gRPC endpoint. The hedera_getNodeAddresses method allows a dApp to request a list of node wallet addresses known to the controller. The controller should only include nodes in this list that it is willing and able to submit transactions to at the time of the request.

Parameters
This method requires no input parameters. The [CAIP-217](https://chainagnostic.org/CAIPs/caip-217) Scope property of the RPC call shall be used to identify which network shall be assumed when obtaining the list of known network nodes.

Returns
nodes – an array of strings formatted in the <shard>.<realm>.<num> format identifying an Hedera Address. The controller should only return known nodes that it can communicate with to submit transactions.

**Does this issue add new functionality to this repository or change functionality (such as a change in function signature)**
This is new functionality

### PR DESCRIPTION
**Pull request by BeeJeeNinja**
• signQueryAndSend - ready closes #14 
• hedera_signMessage - ready closes #15 
• hedera_getNodeAddresses - ready, but still with encoding issues (I guess we need to merge it as is) closes #16

### CODE REVIEW

File: src/examples/typescript/dapp/main.ts [hunk 1]
```diff
@@ -10,15 +10,19 @@ import {
...
+  AccountInfoQuery
 } from '@hashgraph/sdk'
 import {
   HederaChainId,
   HederaSessionEvent,
   HederaJsonRpcMethod,
   transactionToBase64String,
   base64StringToTransaction,
-} from '@hashgraph/walletconnect'
+} from '../../../lib/index'
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: I don't think we need this change?

File: src/examples/typescript/dapp/main.ts [hunk 2]
```diff
@@ -201,3 +209,99 @@ async function disconnect(e: Event) {
   }
 }
 document.querySelector<HTMLFormElement>('#disconnect').onsubmit = disconnect
+
+async function hedera_getNodeAddresses(event) {
+  event.preventDefault();
+
+  const activeSession = signClient.session
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: was there trouble using the globally available session?
if not, let's use that
 https://github.com/hgraph-io/hedera-walletconnect-utils/blob/9fe6d5a85ef355c42ecc6e918d83ca8395eb3231/src/examples/typescript/dapp/main.ts#L29

File: src/examples/typescript/dapp/main.ts [hunk 3]
```diff
@@ -201,3 +209,99 @@ async function disconnect(e: Event) {
...
+    .reverse()
+    .find((session: { expiry: number }) => session.expiry > Date.now() / 1000)
+
+  const messages: string[] = [];
+  if (state['message-to-sign-1']) messages.push(state['message-to-sign-1']);
+  if (state['message-to-sign-2']) messages.push(state['message-to-sign-2']);
+
+  if (!messages.length) {
+    throw Error('There is no messages to sign');
+  }
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: let's just add a required field in the html for both of these messages, then we can reduce the number of lines of code someone has to go through to see the steps

File: src/examples/typescript/dapp/main.ts [hunk 4]
```diff
@@ -201,3 +209,99 @@ async function disconnect(e: Event) {
...
+
+  if (queryName === 'AccountInfoQuery') {
+    query = new AccountInfoQuery()
+      .setAccountId(state['query-account-id']);
+  } else if (queryName === 'FileContentsQuery') {
+    query = new FileContentsQuery()
+      .setFileId(FileId.fromString(state['query-file-id']));
+  }
+
+  const base64Query = Buffer.from(query.toBytes()).toString('base64');
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: great for this PR. we can move this to a helper in the lib just like we have for transaction

File: src/lib/wallet/index.ts [hunk 5]
```diff
@@ -64,7 +66,7 @@ export default class Wallet extends Web3Wallet implements HederaNativeWallet {
     privateKey: string,
     _provider?: Provider,
   ): HederaWallet {
-    const network = chainId.split(':')[1]
+    const network = chainId.split(':')[1];
```
Comment (hgraphql) [inline src/lib/wallet/index.ts]: added this ticket regarding prettier rules - https://github.com/orgs/hgraph-io/projects/4?pane=issue&itemId=44613969

File: src/examples/typescript/dapp/main.ts [hunk 6]
```diff
@@ -201,3 +209,99 @@ async function disconnect(e: Event) {
...
+
+document.getElementById('hedera_getNodeAddresses').onsubmit = hedera_getNodeAddresses
+
+async function hedera_signMessage(event) {
+  const state = saveState(event);
+
+  const activeSession = signClient.session
+    .getAll()
+    .reverse()
+    .find((session: { expiry: number }) => session.expiry > Date.now() / 1000)
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: same here, if we can use the activeSession as defined below, let's do that

File: src/examples/typescript/dapp/main.ts [hunk 7]
```diff
@@ -201,3 +209,99 @@ async function disconnect(e: Event) {
...
+document.getElementById('hedera_signMessage').onsubmit = hedera_signMessage;
+
+
+async function hedera_signQueryAndSend(event, queryName: string) {
+  const state = saveState(event);
+
+  const activeSession = signClient.session
+    .getAll()
+    .reverse()
+    .find((session: { expiry: number }) => session.expiry > Date.now() / 1000)
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: activeSession as defined above

File: src/examples/typescript/wallet/main.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // https://github.com/WalletConnect/walletconnect-monorepo/tree/v2.0/packages/web3wallet
 import type { Web3WalletTypes } from '@walletconnect/web3wallet'
 import { getSdkError } from '@walletconnect/utils'
-import { Wallet, HederaChainId } from '@hashgraph/walletconnect'
+import { Wallet, HederaChainId } from '../../../lib'
```
Comment (hgraphql) [inline src/examples/typescript/wallet/main.ts]: do we need to make this change?

File: src/examples/typescript/wallet/main.ts [hunk 9]
```diff
@@ -46,25 +46,15 @@ async function init(e: Event) {
...
-        `Do you want to proceed with this transaction?: ${JSON.stringify({
-          network: chainId.split(':')[1],
-          accountId,
-          method,
-          body,
-        })}`,
-      )
-    )
-      throw getSdkError('USER_REJECTED_METHODS')
+    const { chainId } = wallet.parseSessionRequest(event)
```
Comment (hgraphql) [inline src/examples/typescript/wallet/main.ts]: we need an end user to confirm whether or not they will accept the proposed transaction

File: src/lib/wallet/index.ts [hunk 10]
```diff
@@ -105,33 +107,46 @@ export default class Wallet extends Web3Wallet implements HederaNativeWallet {
...
     }
   }
 
+
   async executeSessionRequest(
     event: Web3WalletTypes.SessionRequest,
     hederaWallet: HederaWallet,
   ): Promise<void> {
-    const { id, topic, method, body } = this.parseSessionRequest(event)
+    const { id, topic, method, body } = this.parseSessionRequest(event);
```
Comment (hgraphql) [inline src/lib/wallet/index.ts]: no need to make this change

**Review by hgraphql — CHANGES_REQUESTED** [review]
relatively minor changes requested for consistency

File: src/examples/typescript/dapp/main.ts (hunk 1 above)
Comment (BeeJeeNinja) [reply src/examples/typescript/dapp/main.ts]: no, we don't. I have removed it, but I still have build issues. I will try to deal with it myself

**Comment by BeeJeeNinja** [comment]
pushed a new commit, please check

File: src/examples/typescript/dapp/main.ts [hunk 11]
```diff
@@ -10,15 +10,21 @@ import {
...
+  FileContentsQuery,
+  FileId,
+  AccountInfoQuery
 } from '@hashgraph/sdk'
 import {
   HederaChainId,
   HederaSessionEvent,
   HederaJsonRpcMethod,
   transactionToBase64String,
+  queryToBase64String,
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts:20]: nice!

File: src/examples/typescript/dapp/main.ts [hunk 12]
```diff
@@ -201,3 +211,94 @@ async function disconnect(e: Event) {
...
 }
 document.querySelector<HTMLFormElement>('#disconnect').onsubmit = disconnect
+
+async function hedera_getNodeAddresses(event) {
+  event.preventDefault();
+
+  activeSession = signClient.session
+    .getAll()
+    .reverse()
+    .find((session: { expiry: number }) => session.expiry > Date.now() / 1000)
```
Comment (hgraphql) [inline src/examples/typescript/dapp/main.ts]: do we need to redefine this?

File: src/lib/wallet/index.ts [hunk 13]
```diff
@@ -105,33 +108,60 @@ export default class Wallet extends Web3Wallet implements HederaNativeWallet {
...
     const {
-      request: { method, params },
+      request: { method, params: requestParams },
       chainId,
     } = event.params
+    
+
+    let transaction: Transaction | undefined;
+
+    const methodsWithoutTransaction = [
```
Comment (hgraphql) [inline src/lib/wallet/index.ts]: should be named methodsWithTransaction

File: src/lib/wallet/index.ts [hunk 14]
```diff
@@ -105,33 +108,60 @@ export default class Wallet extends Web3Wallet implements HederaNativeWallet {
...
+      HederaJsonRpcMethod.SignTransactionAndSend,
+    ];
+
+    if (methodsWithoutTransaction.includes(method as HederaJsonRpcMethod)) {
+      transaction = base64StringToTransaction(requestParams[0]);
+    }
+
+    const accountId = transaction?.transactionId?.accountId!;
+
+    let query: Query<any> | undefined;
```
Comment (hgraphql) [inline src/lib/wallet/index.ts]: nice!  I noticed query required a type parameter as well...in contrast to `Transaction`

File: src/lib/wallet/index.ts [hunk 15]
```diff
@@ -105,33 +108,60 @@ export default class Wallet extends Web3Wallet implements HederaNativeWallet {
...
   }
 
+
   async executeSessionRequest(
     event: Web3WalletTypes.SessionRequest,
     hederaWallet: HederaWallet,
   ): Promise<void> {
-    const { id, topic, method, body } = this.parseSessionRequest(event)
+    const { id, topic, method, transaction, query, requestParams } = this.parseSessionRequest(event);
+    const body = transaction ?? query ?? requestParams;
```
Comment (hgraphql) [inline src/lib/wallet/index.ts]: I'm not sure if we want to conflate params with body here

**Review by hgraphql — DISMISSED** [review]
lgtm, left a few additional comments for further discussion

**Review by hgraphql — APPROVED** [review]
adding functionality for 3 methods

**Pull request merged**